### PR TITLE
ansible: replace DO debian9-x64-1 with debian12-x64-1

### PR DESCRIPTION
**Pull request by targos**
Refs: https://github.com/nodejs/build/issues/3499

### CODE REVIEW

**Comment by targos** [comment]
Draft because the Python setup changed from previous versions and I don't know how to install `tap2junit` with it.

```
  cmd: [
    '/usr/bin/python3',
    '-m',
    'pip.__main__',
    'install',
    'tap2junit==0.1.6'
  ],

:stderr: error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.

    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.

    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.

    See /usr/share/doc/python3.11/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```

**Comment by cclauss** [comment]
`usr/bin/env python3 -m pip install tap2junit==0.1.6`

The `.__main__` thing seems like something to delete.

**Comment by targos** [comment]
`/usr/bin/env python3 -m pip install tap2junit==0.1.6` prints the same error.

Would it be ok to install with `--break-system-packages` ?

File: ansible/roles/baselayout/vars/main.yml [hunk 1]
```diff
@@ -75,6 +75,10 @@ packages: {
     'gcc-10,g++-10,ccache,git,curl,libfontconfig1,apt-transport-https,ca-certificates,sudo,python3-pip',
   ],
 
+  debian12: [
+    'gcc-12,g++-12,ccache,git,curl,libfontconfig1,apt-transport-https,ca-certificates,sudo,python3-pip',
```
[the remark this answers is not in the document]
Comment (cclauss) [reply ansible/roles/baselayout/vars/main.yml:79]: I think `python3 -m pip install --user tap2junit==0.1.6` is much safer than `--break-system-packages`

[the remark this answers is not in the document]
Comment (targos) [reply ansible/roles/baselayout/vars/main.yml:79]: The ansible config used to install tap2junit is https://github.com/nodejs/build/blob/main/ansible/roles/jenkins-worker/tasks/partials/tap2junit/pip.yml

I'm not sure how to change it so it works.

**Comment by targos** [comment]
With the last commit, `tap2junit` is installed in a virtualenv:

```
# ls /home/iojs/venv/bin
activate      activate.fish  activate.ps1      pip	 pip3	  python   python3.11  wheel	   wheel3
activate.csh  activate.nu    activate_this.py  pip-3.11  pip3.11  python3  tap2junit   wheel-3.11  wheel3.11
```

Now what's the best way to have it activated when jenkins runs?

**Comment by targos** [comment]
`/home/iojs/venv/bin` is a Python script with hashbang: `#!/home/iojs/venv/bin/python`

File: ansible/roles/baselayout/vars/main.yml [hunk 2]
```diff
@@ -75,6 +75,10 @@ packages: {
     'gcc-10,g++-10,ccache,git,curl,libfontconfig1,apt-transport-https,ca-certificates,sudo,python3-pip',
   ],
 
+  debian12: [
+    'gcc-12,g++-12,ccache,git,curl,libfontconfig1,apt-transport-https,ca-certificates,sudo,python3-pip,python3-virtualenv',
```
Comment (cclauss) [inline ansible/roles/baselayout/vars/main.yml:79]: Can we use Python's built-in [`venv`](https://docs.python.org/3/library/venv.html) via `python3-venv` and `python3-full` instead?

Comment (targos) [reply ansible/roles/baselayout/vars/main.yml:79]: Sure, let me try.

Comment (targos) [reply ansible/roles/baselayout/vars/main.yml:79]: What's the command to run in this case? Ansible defaults to `virtualenv` which doesn't exist with `python3-venv`: https://docs.ansible.com/ansible/latest/collections/ansible/builtin/pip_module.html#parameter-virtualenv_command

Comment (targos) [reply ansible/roles/baselayout/vars/main.yml]: I went with `virtualenv_command: python3 -m venv`

**Comment by targos** [comment]
Updated to use `venv` instead of `virtualenv`. Question above is still open.

File: ansible/roles/baselayout/vars/main.yml (hunk 2 above)
Comment (cclauss) [reply ansible/roles/baselayout/vars/main.yml]: Install the venv: `python3 -m venv name_of_my_venv`
Activate the venv (optional): `source name_of_my_venv/bin/activate`

**Comment by targos** [comment]
I can think of solutions but can't say if they're good:
- Add the location of the venv/bin directory to the PATH in https://github.com/nodejs/build/blob/main/ansible/roles/jenkins-worker/templates/systemd.service.j2#L21
- Add a step that creates a symlink of `tap2junit` in `/usr/bin` or `/usr/local/bin`
- Source `venv/bin/activate` if it exists before starting Jenkins

**Comment by cclauss** [comment]
* [x] Add the location of the venv/bin directory to the PATH

**Comment by targos** [comment]
Subsidiary question: at the start or the end of the PATH?

Another question: the current setup installs the venv as root, so maybe we want to do it in another place than `/home/iojs`, or we want to install it as the `iojs` user?

**Comment by cclauss** [comment]
https://packages.debian.org/stable/pipx also brings along `python3-venv` and friends.

**Comment by targos** [comment]
Running with `pipx` sounds much more involved because we would need to do it conditionally until all CI hosts are setup with it.

**Review by cclauss — APPROVED** [review]

**Comment by targos** [comment]
Test build: https://ci.nodejs.org/job/node-test-commit-linux-targos/nodes=debian12-x64/1/

**Comment by targos** [comment]
https://ci.nodejs.org/job/node-test-commit-linux-targos/3/nodes=debian12-x64/

**Comment by targos** [comment]
Test run is green.

**Review by richardlau — APPROVED** [review]

**Pull request merged**